### PR TITLE
Support for navigating panes by MRU

### LINKED ISSUE #2871: Feature Request: Add nextPane and prevPane keybindings
There's `nextTab` and `prevTab`, but since I work in a single-tab, multi-pane environment, those keybindings do nothing for me. How about being able to bind `ctrl+tab` to `nextTab`?

(Would fit in nicely with the rest of #1000 .)

### PR DESCRIPTION
**Pull request by PankajBhojwani**
Adds a "move to previous pane" and "move to next pane" keybinding, which
navigates to the last/first focused pane

We assign pane IDs on creation and maintain a vector of active pane IDs
in MRU order. Navigating panes by MRU then requires specifying which
pane ID we want to focus. 

From our offline discussion (thanks @zadjii-msft for the concise
description):

> For the record, the full spec I'm imagining is:
> 
> { command": { "action": "focus(Next|Prev)Pane", "order": "inOrder"|"mru", "useSwitcher": true|false } },
> 
> and order defaults to mru, and useSwitcher will default to true, when
> there is a switcher. So 
> 
> { command": { "action": "focusNextPane" } },
> { command": { "action": "focusNextPane", "order": "mru" } },
> 
> these are the same action. (but right now we don't support the order
> param)
>  
> Then there'll be another PR for "focusPane(target=id)"
> 
> Then a third PR for "focus(Next|Prev)Pane(order=inOrder)"

> for the record, I prefer this approach over the "one action to rule
> them all" version with both target and order/direction as params,
> because I don't like the confusion of what happens if there's both
> target and order/direction provided. 

References #1000 
Closes #2871

### CODE REVIEW

File: src/cascadia/TerminalApp/Pane.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp:1579]: Can we change this to `PaneId()` and `PaneId(<id>)`?

File: src/cascadia/TerminalSettingsModel/defaults.json
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/defaults.json:335]: I feel like we're using `alt` as the button for "this has to do with panes". Maybe these should be `ctrl+alt+left/right`?

File: src/cascadia/TerminalApp/Pane.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp:1581]:
```suggestion
void Pane::FocusPane(const uint16_t id)
```
and add a comment saying what the param is?

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
A few other things:
- [ ] Don't forget to update the docs/schema
- [ ] what happens if we provide "next" or "previous" to "resizePane"?

**Comment by PankajBhojwani** [comment]
> what happens if we provide "next" or "previous" to "resizePane"?

The `_Resize()` function will return false. I've noted in the docs that the direction given to `resize` cannot be next or previous. If people feel strongly about it we can make them two distinct direction enums?

File: doc/cascadia/profiles.schema.json
Comment (carlos-zamora) [inline doc/cascadia/profiles.schema.json:283]: But `next` and `previous` are still displayed by the schema. I'm worried we'll get a few bug reports on that. So I recommend we definitely split the schema into enums.

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
> The `_Resize()` function will return false and nothing happens. I've noted in the docs that the direction given to `resize` cannot be next or previous. If people feel strongly about it we can make them two distinct direction enums? 

A few issues here:
1. Generally, when you put an invalid value for the arg in the json, we display a warning.
2. When you say "nothing happens" does it eat the keybinding or does the keybinding still get sent to the buffer (i.e. `^A`)?
3. Right now, we have enum mappers exposing enums to the Settings UI. Though keybindings aren't there yet, we'll probably use this the same way. If we have a single enum mapper for `Direction`, we'll have to figure out how to exclude "next" and "previous" in one keybinding arg, but not the other.

I'd rather you split the enum into 2 different ones in the schema and the code. It just acts more properly that way and it's more future-proof.

**Comment by PankajBhojwani** [comment]
> I'd rather you split the enum into 2 different ones in the schema and the code. It just acts more properly that way and it's more future-proof.

Fair enough, its split now!

**Review by carlos-zamora — APPROVED** [review]
Thank you. You _could_ keep the `ResizeDirection` as just plain `Direction` if you wanted to reduce the diff. But I'm fine either way.

**Comment by zadjii-msft** [comment]
Okay my immediate thoughts are around how we expose this action to the user. We're already running into similar struggles with `next/prevTab` and the tab switcher, so maybe we should try and pre-empt those issues with pane switching?

For reference, there's also discussion in #8025 that might be helpful.

So, with next/prev MRU switching, without some sort of dialog, those will just toggle between the last two used panes. If there's a fourth pane, that pane will be inaccessible with next/prev pane.

Also, having panes in the ATS is something we might want some day, with something like:

> * Tab 1:
>   - **Pane 1: cmd.exe**
>   - Pane 2: pwsh.exe
> * Tab 2:
>   - Pane 1: wsl.exe

whatever, you get it.

Plus, do we want to enable in-order next/prev traversal? Once we have pane IDs, then presumably, yea. 
<hr>

Do we want to overload `moveFocus` to accept `next` and `prev`, or would having another action entirely be better? Here are some proposals:

```jsonc
// Focus pane 1
// - This is sensible, no arguments here
{ command": { "action": "focusPane", "id": 1 } },

// Focus the next in order pane, don't use the switcher
// - We'd have to ensure that an id and order aren't passed at the same time, and that might be unintuitive
// - How do we specify next/prev? "direction" seems wrong, but maybe we keep the
//   FocusDirection(up|down|left|right|next|prev) enum and have one focus action 
//   to rule them all
{ command": { "action": "focusPane", "order": "inOrder", useSwitcher": false } },

// Focus the next MRU pane
// - Withough the switcher, this can only go one pane deep in the MRU stack
// - presumably once there's a pane switcher, it would default to enabled?
{ command": { "action": "focusNextPane", "order": "mru" } },

// Focus the prev inOrder pane
// - this seems straightforward
{ command": { "action": "focusPrevPane", "order": "inOrder" } },

// Focus the next pane, in mru order, explicitly disable the switcher
// - The user opted in to only being able to MRU switch one deep. That's fine, that's what they want.
{ command": { "action": "focusNextPane", "order": "mru", "useSwitcher": false} },

// Focus the prev inOrder pane, explicitly with the switcher
// - Maybe they disabled the switcher globally, but what it on for this action? 
{ command": { "action": "focusPrevPane", "order": "inOrder", "useSwitcher": true } },
```

Thoughts?

**Comment by zadjii-msft** [comment]
@msftbot make sure I sign off on this

**Comment by zadjii-msft** [comment]
@msftbot make sure @zadjii-msft signs off on this

**Comment by PankajBhojwani** [comment]
I like the idea of having separate commands for `focusPrevPane` and `focusNextPane` in preparation for the pane switcher. 

> Plus, do we want to enable in-order next/prev traversal? Once we have pane IDs, then presumably, yea.

I am not sure about this, we will need a way for the user to know the pane IDs and that may not always be feasible - paging @DHowett for thoughts on this

**Comment by carlos-zamora** [comment]
> Do we want to overload `moveFocus` to accept `next` and `prev`, or would having another action entirely be better? Here are some proposals:

@zadjii-msft @PankajBhojwani Could we create a spec for this and discuss the design in a brownbag? There seems to be a lot of options here. And I'd like to make sure the design interacts well with the current ATS customization model. Especially given our recent experience with ATS MRU.

**Comment by DHowett** [comment]
I don’t think this requires a spec. An informal chat should do!

**Comment by cinnamon-msft** [comment]
Is `{ command": { "action": "focusPrevPane", "order": "inOrder" } },` in order of creation or in order from left-right, top-bottom? Also does this order change once we get pane rearrangement?

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalTab.cpp:285]: So the Tab is responsible for assigning IDs here as panes get split, that makes sense.

However, what happens when I make two panes `[ 1 | 2 ]`, then close pane 1. The parent pane will absorb the control from pane 2 into itself, but won't that pane still have the ID `1`? 

Should the Pane make sure to clear out it's assigned ID when a split is being made, and also make sure to absorb the ID of the remaining child (if needed)?

File: src/cascadia/TerminalApp/Pane.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Pane.cpp:1573]: Should this just be `Id`? (`PaneId` seems like the `Pane` bit is redundant)

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I'm mostly blocking over the pane closing, ID transferring thing.

While Pane ID's aren't addressable now, we should consider how we're handling the incrementing of pane IDs, probably sooner than later (though maybe in a follow-up PR). 
  * I know that `tmux` assigns IDs based on positions, though I kinda hate that. 
  * Right now we're assigning a single static ID to a pane, and it will never change. But this also means that closing a pane 3 in panes [1, 2, 3, 4] would leave panes with IDs [1, 2, 4], and the next pane would be assigned 5 as it's ID.
    - Perhaps that would be less weird if the next pane was assigned an ID of 3? Perhaps that would be more weird? idk.
  * There's also the potential that closing a pane _x_ would decrement the ID's of all panes [_x+1_, _N_) (where _N_=number of panes), so that no gaps remain. So if we closed pane 2 of [1, 2, 3, 4], we'd end up with panes with IDs [1, \_, 2, 3]

Those are all showerthoughts, something to consider for the future.

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/ActionArgs.cpp:183]: Oh hey, we need to add a resource for the name of this action as well

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (PankajBhojwani) [reply src/cascadia/TerminalApp/TerminalTab.cpp:522]: `weakPane.get()` causes a "cannot access protected member" error

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (PankajBhojwani) [reply src/cascadia/TerminalApp/TerminalTab.cpp:349]: I don't think so - we only get to this point if it even is possible to move focus (i.e. there's more than 1 leaf pane) and so there will be at least 2 entries in `_mruPanes`

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (PankajBhojwani) [reply src/cascadia/TerminalApp/TerminalTab.cpp]: Thanks for the catch! We absorb the remaining child's ID and clear out ID when splitting now

File: src/cascadia/TerminalApp/Pane.h
Comment (DHowett) [reply src/cascadia/TerminalApp/Pane.h:171]: _hmm_. possibly.

Comment (DHowett) [reply src/cascadia/TerminalApp/Pane.h:171]: This might actually work as written

Comment (PankajBhojwani) [reply src/cascadia/TerminalApp/Pane.h]: It does! Pushed the change

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalTab.cpp:282]: perhaps a useful comment:

```suggestion
        // Make sure to take the ID before calling Split() - it will clear out the active pane's ID 
        const auto activePaneId = _activePane->Id();
```

**Comment by zadjii-msft** [comment]
@msftbot don't merge this until 5pm tomorrow

**Comment by zadjii-msft** [comment]
@msftbot merge this in 19 hours

**Review by zadjii-msft — APPROVED** [review]
I think I follow all this, and it looks good to me. I'm telling the bot to wait until EOD Friday to merge this, to give Dustin one last chance, but I think this is fine

File: src/cascadia/TerminalApp/TerminalTab.h
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.h:86]: Does this need to be a whole WinRT vector? Is an `std::vector` sufficient for this since it won't be sent around in any winRT APIs?

Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalTab.h:86]: (a winrt vector is somewhat more expensive than a normal vector, since it's a whole heap-allocated object that can be reference-counted and stuff)

Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalTab.h:86]: (and all functions that operate on it do so through virtual function dispatch)

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionArgs.cpp:185]: wait, don't we need to add a command name string for `DirectionPrevious`?

File: src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:221]: the construction of this sentence will be wrong in most languages.

There is no general sentence structure that can support "Move focus left" and "Move focus previous" at the same time. Especially not in another language, right?

This is different from how we handled other commands, where we said (something like) `Move focus: direction=previous`. That looks like a straight parameter substitution.

This is different . . . it needs its own complete command string, "Move focus to last-used pane" or something.

File: src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:348]: SURPRISED THIS COMPILES
should be 4

File: src/cascadia/TerminalSettingsModel/defaults.json
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/defaults.json:334]: hmm alright. i'm less sure.

make sure you add it to defaults-universal.json!

**Review by DHowett — CHANGES_REQUESTED** [review]

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
Hit this with a code format!

**Pull request merged**
